File: archives/leetcode2/0103. Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: 'TreeNode') -> 'List[List[int]]':
        # Definition for a binary tree node.
        if not root:
            return []
        root_stack=[root]
        rep=[]
        flag=0
        while root_stack:
            len_root_stack=len(root_stack)
            tmp=[]
            for _ in range(len_root_stack):
                cur_root=root_stack.pop(0)
                tmp.append(cur_root.val)
                if cur_root.left:
                    root_stack.append(cur_root.left)
                if cur_root.right:
                    root_stack.append(cur_root.right)
            if flag==0:
                rep.append(tmp)
                flag=1
            else:
                rep.append(tmp[::-1])
                flag=0
        return rep

File: archives/C#/0103. Binary Tree Zigzag Level Order Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
        IList<IList<int>> rep=new List<IList<int>>();
        if(root==null){return rep;}
        Queue<TreeNode> rootQueue=new Queue<TreeNode>();
        rootQueue.Enqueue(root);
        bool flag=true;
        while(rootQueue.Count()!=0){
            int lenRootQueue=rootQueue.Count();
            IList<int> rootList=new List<int>();
            for(int i=0;i<lenRootQueue;i++){
                TreeNode curRoot=rootQueue.Dequeue();
                rootList.Add(curRoot.val);
                if(curRoot.left!=null){
                    rootQueue.Enqueue(curRoot.left);
                }
                if(curRoot.right!=null){
                    rootQueue.Enqueue(curRoot.right);
                }
            }
            if(flag==true){
                rep.Add(rootList);
                flag=false;
            }
            else{
                List<int> tmp=new List<int>(rootList);
                tmp.Reverse();
                IList<int> tmp1=new List<int>(tmp);
                rep.Add(tmp1);
                flag=true;
            }
        }
        return rep;
    }
}
